crates/astria-conductor/src/conductor/inner.rs

```diff
@@ -108,28 +127,34 @@ impl Inner {
         self.restart_or_shutdown(exit_reason).await
     }
 
-    /// Shuts down all tasks.
+    /// Shuts down all tasks and returns a token indicating whether to restart or not.
     ///
     /// Waits 25 seconds for all tasks to shut down before aborting them. 25 seconds
     /// because kubernetes issues SIGKILL 30 seconds after SIGTERM, giving 5 seconds
     /// to abort the remaining tasks.
     #[instrument(skip_all, err, ret(Display))]
     async fn restart_or_shutdown(
         mut self,
-        exit_reason: eyre::Result<ShutdownSignalReceived>,
+        exit_reason: eyre::Result<ConductorExitReason>,
     ) -> eyre::Result<RestartOrShutdown> {
-        self.shutdown_token.cancel();
```

Comment:
perform the cancellation here, but have `conductor::Inner` track the "public" shutdown token and a "private" child token, which gets also passed to the tasks. Then you can always cancel the child token but still check if the public token was cancelled.